File: android/src/main/java/expo/modules/applemusic/AppleMusicRestStack.kt
package expo.modules.applemusic

import android.content.Context

/** Wires shared REST transport + domain clients for Android services. */
internal class AppleMusicRestStack(context: Context) {
  val transport: AppleMusicRestTransport = OkHttpAppleMusicRestTransport(context)
  val storefront: StorefrontRestClient = StorefrontRestClient(transport)
  val catalog: CatalogRestClient = CatalogRestClient(transport, storefront)
  val library: LibraryRestClient = LibraryRestClient(transport)
  val history: HistoryRestClient = HistoryRestClient(transport)
  val ratings: RatingsRestClient = RatingsRestClient(transport)
  val libraryMutations: LibraryMutationsRestClient = LibraryMutationsRestClient(transport)
  val recommendations: RecommendationsRestClient = RecommendationsRestClient(transport)

  companion object {
    fun create(context: Context): AppleMusicRestStack = AppleMusicRestStack(context)
  }
}

File: src/rest/apple-music-rest-stack.ts
import { CatalogRestClient } from './catalog-rest-client';
import { HistoryRestClient } from './history-rest-client';
import { LibraryMutationsRestClient } from './library-mutations-rest-client';
import { LibraryRestClient } from './library-rest-client';
import { RatingsRestClient } from './ratings-rest-client';
import { RecommendationsRestClient } from './recommendations-rest-client';
import { StorefrontRestClient } from './storefront-rest-client';
import type { AppleMusicRestTransport } from './apple-music-rest-transport';
import type { StorefrontResolver } from './storefront-rest-client';

export type AppleMusicRestStack = {
  transport: AppleMusicRestTransport;
  storefront: StorefrontRestClient;
  catalog: CatalogRestClient;
  library: LibraryRestClient;
  history: HistoryRestClient;
  ratings: RatingsRestClient;
  libraryMutations: LibraryMutationsRestClient;
  recommendations: RecommendationsRestClient;
};

export function createAppleMusicRestStack(
  transport: AppleMusicRestTransport,
  storefrontResolver?: StorefrontResolver,
): AppleMusicRestStack {
  const storefront = new StorefrontRestClient(transport, storefrontResolver);
  return {
    transport,
    storefront,
    catalog: new CatalogRestClient(transport, storefront),
    library: new LibraryRestClient(transport),
    history: new HistoryRestClient(transport),
    ratings: new RatingsRestClient(transport),
    libraryMutations: new LibraryMutationsRestClient(transport),
    recommendations: new RecommendationsRestClient(transport),
  };
}
